streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1530,6 +1537,24 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
    */
   final def runDrain: ZIO[R, E, Unit] = run(Sink.drain)
 
+  final def schedule[R1 <: R, E1 >: E, B](schedule: ZSchedule[R1, A, B]): ZStream[R1 with Clock, E1, B] =
```

Comment:
This is really nice. Can it also replace `spaced`?

Reply:
Yeah I'll try to 👍

Reply:
There's some weird behaviour when I use a schedule, where the second stream in my test makes progress much faster than the first one in the init phase. I will investigate.

Reply:
Found it. The schedule will now contribute to deciding the end of the stream. So, if the schedule ends after 2 repeats, only 3 elements will be produced. It seems more intuitive to  me (see the definition of `fixed`), but I'm keen to hear your opinion. If ok I'll move on to look at `spaced`.

Reply:
I actually found the previous behavior, in which `decision.cont == false` means that the schedule is restarted, to be more intuitive. This is also more in line with `aggregateWithin`, which restarts the schedule when it stops.

Could you explain what's the advantage?

By the way, we can attain this behavior with the previous formulation if you emit the schedule's output into the stream. We can then use `takeWhile` or `collectWhile` to end the stream when the schedule ends.

Reply:
I don't know if there's an advantage, I guess it's more that I got confused between the name and the intent. `spaced` has different meanings between schedule and streams and that's where I got confused: in the former it means there's a fixed duration between each cycle, in the latter it can mean any kind of schedule, even ones that don't wait.

Reply:
Ok, I'm fairly certain that the previous formulation is more correct and we'd want to distribute the schedule's effect repeatedly over the stream rather than have it short-circuit the stream.

If you restart the schedule when it ends, that formulation lets you use `Schedule.recurs(2)` to repeat each element twice.

Reply:
Right 🤦‍♂ so I literally woke up during the night with a solution following your suggestion with `collectAll`:
- this function will be `scheduleWith`, emitting both the stream and schedule's output with two functions for aligning types
- `scheduleEither` will map to either (and replace the current `spacedEither`)
- I am not sure about my `fixed` and the existing `spaced`, there seems to be a semantic clash, but in any case the `fixed` is just `scheduleEither(...).collectAll(case Left(x) => x)` indeed.

Reply:
Hehe, I had a realization myself. There are two ways to apply a schedule to a stream with regards to the schedule's `cont`.

Given a stream `A B C` and a schedule `(Schedule.exponential(2.millis) && Schedule.recur(2) && Schedule.const("Done")).map(_._2)`, these are the interpretations:
1. As long as the schedule wants to continue, repeat the current element after a delay:
`A |-2ms-| A |-4ms-| A |-8ms-| "Done" B |-2ms-| B |-4ms-| B |-8ms-| "Done" ...`

2. As long as the schedule wants to continue, emit the stream's next element after a delay:
`A |-2ms-| B |-4ms-| C |-8ms-| "Done"`

The first interpretation is the existing `spaced`. I guess that this implementation should be the second interpretation.

Reply:
dang, I was quite happy with the terseness of `scheduleWith`. there may be some confusion as to which does what, are you ok with naming the one that acts on each elements `scheduleElementsWith` and the one that acts on the whole stream `scheduleWith` (I also think it would fix the confusion between `ZStream.spaced` and `ZSchedule.spaced`)?

Reply:
Yeah, those names are good! We can bring `ZStream#repeat` inline with that too:
- The signature can be changed to:
```scala
final def repeat[R1 <: R, B](schedule: ZSchedule[R1, Any, B]): ZStream[R1 with Clock, E, Either[B, A]]
```
- And a `repeatWith` variant can be added.

Can be done in a follow-up of course :-)

Reply:
Let's do in a follow-up, I'd like to see where the ticket on combinators is going.

Reply:
Ending with the schedule is more "foundational", since you can turn a terminating schedule into a forever schedule with `Schedule#forever`. However it is less convenient. Usually this argues for two methods: one to do the more essential thing, and the other, more visible and obvious one, to do the `Schedule#forever` thing.